### ISSUE TITLE
Get rid of heap pseudoconstructors ?

### ISSUE THREAD

**Issue by oxinabox** [JuliaCollections/DataStructures.jl#469]
https://github.com/JuliaCollections/DataStructures.jl/blob/31044efe5130b8d1980e2b6650337da08f27274e/src/heaps/binary_heap.jl#L119-L125

The heap code has these pseudoconstructors, written in `snake_case`
This code is 6 years old and was written just a month after the julia 0.1 release.
I am feeling like they were a work around for some thing 0.1 was missing.
(In general a lot of the code had workarounds for things like the inability to just put type-parameters on functions without using them in the arguments. That is moistly gone now).

I think we can move these pseudoconstructors into overloads of the real constructors (in ProperCase),
and deprecate the old ones.

**Comment by dillondaudert** [comment]
I'm willing to go ahead and give this a shot. How do you think users should instantiate the min/max versions of the heap? The first thing I thought of would be `BinaryHeap{T, Comp}() where {T, Comp} = new{T, Comp}(Vector{T}())` however that would require the user passing `GreaterThan()`/`LessThan()` themselves. 

The `GreaterThan`/`LessThan` types look like a holdover from when `Base.Order` didn't exist, and that seems OK as long as the users aren't forced to interact with them directly. If they are, I would much rather it be `BinaryHeap{T, O}() where {T, O<:Ordering} = new{T, O}(Vector{T}())`. Of course, that would require a more extensive rewrite of the heaps...

**Comment by oxinabox** [comment]
I was thinking more along the lines of creating a type alias style const.

``` 
const BinaryMinHeap{T} = BinaryHeap{T,LessThan}
const BinaryMaxHeap{T} = BinaryHeap{T,GreaterThan}
```
I think that works.


This is not to say that the more extensive rewrite to use `Base.Order` is not needed.

**Comment by dillondaudert** [comment]
Ok. Once #468 is merged, I'll make a PR for this.

**Comment by dillondaudert** [comment]
This is a little tricky due to the ordering of the parametric arguments and how I'm thinking the constructors should work.
I'm going off of desiring two ways to instantiate a heap:
- `Heap{T}()`: initialize empty heap with element types `T`
- `Heap(xs::AbstractVector{T})`: initialize a heap with array of elements of type `T`.

For the second to be possible with `BinarMinHeap`/`BinaryMaxHeap`, the type aliases need to be defined without type parameters. However, if we do `const BinaryMinHeap = BinaryHeap{LessThan}` then

`BinaryMinHeap{Int} -> BinaryHeap{LessThan, Int64}`,

so the parametric args are reversed. We could reverse them in the struct definition, but I'm not sure if that will have any downstream consequences?

**Comment by oxinabox** [comment]
I don't think I understand what you are getting at?
Why does the type alias need to be defined without type parameters?

**Comment by oxinabox** [comment]
It is still a DataType that is being given a call overload i.e. a constructor, not just a function.
(well technically it is a `UnionAll` being given a call overload)

TBH, my main issue was the snake-case.

**Comment by oxinabox** [comment]
I think this was resolved in #472, modulo actually removing them, which will naturally follow from deprecating thenm. 
so I am going to close this
